lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -53,18 +57,25 @@ class ArrayAllocation {
 
   // The calls to Array get_element that use this array allocation.
   llvm::SmallSetVector<ApplyInst *, 16> GetElementCalls;
+  llvm::SmallVector<ApplyInst *, 4> AppendContentsOfCalls;
   llvm::DenseMap<uint64_t, SILValue> ElementValueMap;
+  llvm::SmallVector<SILValue, 4> ElementValueVector;
 
   // Array get_element calls and their matching array element value for later
   // replacement.
-  llvm::SmallVectorImpl<std::pair<ApplyInst *, SILValue>> &ReplacementMap;
+  llvm::SmallVectorImpl<std::pair<ApplyInst *, SILValue>> &GetElementReplacementMap;
+
+  // Array append_contentsOf calls and their matching array values for later
+  // replacement.
+  llvm::SmallVectorImpl<std::pair<ApplyInst *, llvm::SmallVector<SILValue, 4> > > &AppendContentsOfReplacementMap;
 
   ArrayAllocation(
       ApplyInst *AI,
-      llvm::SmallVectorImpl<std::pair<ApplyInst *, SILValue>> &Replacements)
-      : Alloc(AI), ReplacementMap(Replacements) {}
+      llvm::SmallVectorImpl<std::pair<ApplyInst *, SILValue>> &GetElementReplacements,
+      llvm::SmallVectorImpl<std::pair<ApplyInst *, llvm::SmallVector<SILValue, 4> > > &AppendContentsOfReplacements)
```

Comment:
Made the `typedefs`, and removed redundant `llvm` prefixes. I'm not sure what you mean about the helper class, but the refactored code is much cleaner, so maybe it's not necessary anymore.